lib/src/native/realm_core.dart

```diff
@@ -1386,6 +1390,105 @@ class _RealmCore {
     final dynamic profileData = jsonDecode(data.cast<Utf8>().toRealmDartString(freeNativeMemory: true)!);
     return UserProfile(profileData as Map<String, dynamic>);
   }
+
+  SessionHandle realmGetSession(Realm realm) {
+    return SessionHandle._(_realmLib.invokeGetPointer(() => _realmLib.realm_sync_session_get(realm.handle._pointer)));
+  }
+
+  String sessionGetPath(Session session) {
+    return _realmLib.realm_sync_session_get_file_path(session.handle._pointer).cast<Utf8>().toRealmDartString()!;
+  }
+
+  SessionState sessionGetState(Session session) {
+    final value = _realmLib.realm_sync_session_get_state(session.handle._pointer);
+    return _convertCoreSessionState(value);
+  }
+
+  ConnectionState sessionGetConnectionState(Session session) {
+    final value = _realmLib.realm_sync_session_get_connection_state(session.handle._pointer);
+    return ConnectionState.values[value];
+  }
+
+  UserHandle sessionGetUser(Session session) {
+    return UserHandle._(_realmLib.realm_sync_session_get_user(session.handle._pointer));
+  }
+
+  SessionState _convertCoreSessionState(int value) {
+    switch (value) {
+      case 0: // RLM_SYNC_SESSION_STATE_ACTIVE
+      case 1: // RLM_SYNC_SESSION_STATE_DYING
+        return SessionState.active;
+      case 2: // RLM_SYNC_SESSION_STATE_INACTIVE
+      case 3: // RLM_SYNC_SESSION_STATE_WAITING_FOR_ACCESS_TOKEN
+        return SessionState.inactive;
+      default:
+        throw Exception("Unexpected SessionState: $value");
+    }
+  }
+
+  void sessionPause(Session session) {
+    _realmLib.realm_sync_session_pause(session.handle._pointer);
+  }
+
+  void sessionResume(Session session) {
+    _realmLib.realm_sync_session_resume(session.handle._pointer);
+  }
+
+  int sessionRegisterProgressNotifier(Session session, ProgressDirection direction, ProgressMode mode, SessionProgressNotificationsController controller) {
+    final isStreaming = mode == ProgressMode.reportIndefinitely;
+    return _realmLib.realm_dart_sync_session_register_progress_notifier(session.handle._pointer, Pointer.fromFunction(on_sync_progress), direction.index,
+        isStreaming, controller.toPersistentHandle(), _deletePersistentHandleFuncPtr, session.scheduler.handle._pointer);
```

Comment:
Using a persistent handle here, risk us leaking if the user forgets to cancel any subscriptions. This is probably not a big issue for this particular stream.

Reply:
+1. Why use persistent here?

Reply:
I think this will result in behavior that is more intuitive. When using non-streaming progress listeners, we'll automatically unsubscribe whenever we reach the target progress. If the user registers a streaming progress notifier, it's likely they want to get notifications for the lifetime of the Realm, without taking too much care about keeping the stream alive. Unlike object notifications, progress notifications are super cheap, so even if the user forgets to close the stream it's not going to have a perceivable negative impact.